File: 9/904_FruitIntoBaskets.py
# https://leetcode.com/problems/fruit-into-baskets/description/?envType=daily-question&envId=2025-08-04

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        lengthOfFruits=len(fruits)
        left=0
        right=0
        resultantMaxFruits=-1*10000000010
        basketMap={}
        for right in range(0, lengthOfFruits, 1):
            if fruits[right] in basketMap:
                basketMap[fruits[right]]+=1
            else:
                basketMap[fruits[right]]=1
            while(True):
                if len(basketMap)<=2:
                    break
                basketMap[fruits[left]]-=1
                if basketMap[fruits[left]]==0:
                    del(basketMap[fruits[left]])
                left+=1
            resultantMaxFruits=max(resultantMaxFruits, (right-left)+1)
        return resultantMaxFruits

File: 9/904_FruitIntoBaskets.java
/*
https://leetcode.com/problems/fruit-into-baskets/?envType=daily-question&envId=2025-08-04
*/

class Solution {
    public int totalFruit(int[] fruits) {
        int lengthOfFruits=fruits.length, left=0, right=0, resultantMaxFruits=Integer.MIN_VALUE;
        HashMap<Integer, Integer>basketMap=new HashMap<>();
        for(; right<lengthOfFruits; ++right){
            if(basketMap.containsKey(fruits[right])){
                basketMap.put(fruits[right], basketMap.get(fruits[right])+1);
            } else basketMap.put(fruits[right], 1);
            while(true){
                if(basketMap.size()<=2) break;
                basketMap.put(fruits[left], basketMap.get(fruits[left])-1);
                if(basketMap.get(fruits[left])==0){
                    basketMap.remove(fruits[left]);
                }
                left++;
            }
            resultantMaxFruits=Math.max(resultantMaxFruits, (right-left)+1);
        }
        return resultantMaxFruits;
    }
}
